src/backend/telegram/confirmTelegram.ts

```diff
@@ -0,0 +1,111 @@
+import { createHash, createHmac } from 'node:crypto';
+
+import { z } from 'zod';
+import {
+  Request,
+  ResponseObject,
+  ResponseToolkit,
+  ServerRoute,
+} from '@hapi/hapi';
+import Boom from '@hapi/boom';
+
+import { Claim } from '@kiltprotocol/core';
+
+import { getSessionWithDid, setSession } from '../utilities/sessionStorage';
+import { paths } from '../endpoints/paths';
+import { configuration } from '../utilities/configuration';
+
+import { telegramCType } from './telegramCType';
+
+const zodPayload = z.object({
+  sessionId: z.string(),
+  json: z.string(),
+});
+
+export type Input = z.infer<typeof zodPayload>;
+
+export interface Output {
+  id: number;
+  first_name: string;
+  last_name?: string;
+  username?: string;
+}
+
+function validateAuthData(json: string): void {
+  const { hash, ...rest } = JSON.parse(json).auth_data as {
+    hash: string;
+    auth_date: number;
+  };
+
+  const dayAgoTimestamp = Date.now() / 1000 - 24 * 60 * 60;
+  if (rest.auth_date < dayAgoTimestamp) {
+    throw Boom.forbidden('Auth data is too old');
+  }
+
+  const plaintext = Object.entries(rest)
+    .map(([key, value]) => `${key}=${value}`)
+    .sort()
+    .join('\n');
+
+  const secret = createHash('sha256')
+    .update(configuration.telegram.token)
+    .digest();
+
+  const computed = createHmac('sha256', secret)
+    .update(plaintext)
+    .digest()
+    .toString('hex');
+
+  if (hash !== computed) {
+    throw Boom.forbidden('Auth data hash doesn’t match');
+  }
+}
```

Comment:
Maybe someone who understands crypto better than me should review this function 🥴

Reply:
This is described here https://core.telegram.org/widgets/login#checking-authorization, and the check passes, so the hashes are calculated correctly. But we could ask someone else for your peace of mind :)

Reply:
Oh I see. My mind is at peace. Go ahead and merge it 😄

Reply:
Yeah looks good to me too

Reply:
Thank you Raphael!